src/distributions/mod.rs

```diff
@@ -275,42 +276,30 @@ impl<'a, D, R, T> Iterator for DistIter<'a, D, R, T>
 /// let mut rng = thread_rng();
 /// let erased_rng: &mut RngCore = &mut rng;
 /// let val: f32 = erased_rng.sample(Standard);
-/// println!("f32 from (0, 1): {}", val);
+/// println!("f32 from [0, 1): {}", val);
 /// ```
 ///
-/// # Open interval for floats
-/// In theory it is possible to choose between an open interval `(0, 1)`, and
-/// the half-open intervals `[0, 1)` and `(0, 1]`. All can give a distribution
-/// with perfectly uniform intervals. Many libraries in other programming
-/// languages default to the closed-open interval `[0, 1)`. We choose here to go
-/// with *open*, with the arguments:
+/// # Floating point implementation
+/// The floating point implementations for `Standard` generate a random value in
+/// the half-open interval [0, 1).
 ///
-/// - The chance to generate a specific value, like exactly 0.0, is *tiny*. No
-///   (or almost no) sensible code relies on an exact floating-point value to be
-///   generated with a very small chance (1 in 2<sup>23</sup> (approx. 8
-///   million) for `f32`, and 1 in 2<sup>52</sup> for `f64`). What is relied on
-///   is having a uniform distribution and a mean of `0.5`.
-/// - Several common algorithms rely on never seeing the value `0.0` generated,
-///   i.e. they rely on an open interval. For example when the logarithm of the
-///   value is taken, or used as a devisor.
+/// All values that can be generated are multiples of ε/2. For `f32` the 23 most
+/// significant random bits of an `u32` are used, for `f64` 53 from an `u64`.
+/// The conversion uses the common multiply-based approach.
 ///
-/// In other words, the guarantee some value *could* be generated is less useful
-/// than the guarantee some value (`0.0`) is never generated. That makes an open
-/// interval a nicer choice.
+/// The `Open01` distribution provides an alternative: it generates values in
+/// the open interval (0, 1), with one less random bit. It uses the slightly
+/// faster transmute-based approach for the conversion to a floating point
+/// value.
 ///
-/// Consider using `Rng::gen_range` if you really need a half-open interval (as
-/// the ranges use a half-open interval). It has the same performance. Example:
-///
-/// ```
-/// use rand::{thread_rng, Rng};
+/// `Rng::gen_range(0, 1)` also uses the faster transmute-based method, but
+/// produces values in a half-open interval just like `Standard`.
 ///
-/// let mut rng = thread_rng();
-/// let val = rng.gen_range(0.0f32, 1.0);
-/// println!("f32 from [0, 1): {}", val);
-/// ```
+/// As a final alternative, if you require the floating point value to never be
+/// exactly `0.0`: consider using `1.0 - rng.gen()`, which transforms the
+/// half-open `Standard` distribution to the `(0, 1]` interval.
```

Comment:
I would prefer to assume any users who care know what they want:

> If you wish to sample from the `(0, 1]` half-open interval consider using `1.0 - rng.gen()`.

Looking at the code, I couldn't actually think of an alternative which samples with this precision and doesn't use an extra op (though of course one could add `ε/2` or 1 before conversion to float).

Reply:
> Looking at the code, I couldn't actually think of an alternative which samples with this precision and doesn't use an extra op (though of course one could add ε/2 or 1 before conversion to float).

Explicit `OpenClosed01` does have advantages though:
* the intent is clearer and the code less cluttered,
* the `+1` integer addition before conversion is faster than the `1.0 - u` user-side workaround

Reply:
I agree with you. There is also some benefit to an explicit `ClosedOpen01`, even if just for local documentation *(this value is strictly less than 1)*.

Reply:
Yes, this is also something that hit me when looking at the diff: if not for @pitdicker's perceptivity, subtle bugs would have been introduced in Ziggurat and Gamma where `rng.gen()` was previously used. When I read `rng.gen()` it is always unclear to me if the author of the code did so for the convenient notation or whether the code actually relies on the guarantees regarding the `0.0` and `1.0` bounds.

Reply:
I suppose that makes for a good argument.

In this case I had it easy, as I also was the one introducing the ambiguity in https://github.com/rust-lang-nursery/rand/pull/274 :smile:.